### ISSUE TITLE
Security Vulnerability in tf-encrypted Secure Random Module

### ISSUE THREAD

**Issue by GuopengLin** [tf-encrypted/tf-encrypted#900]
## Description:
I have identified a critical security vulnerability in the random number generation module in tf-encrypted.

When the types of seed, minval, and maxval differ, the module generates a constant value instead of random numbers.

## Example:
In the following example, the result (res) is always a constant tensor of [-128, -128] (even the seed is changed), instead of a uniformly random tensor within the specified range.

```
import tensorflow as tf
from tf_encrypted.operations.secure_random.secure_random import secure_random_module

seed = tf.constant([-27587587, -1632666452, -218620675, -1537614956, 1554788191, -1805869028, -74859994, -1320420251], dtype=tf.int32)
shape = (2,)
minval = tf.constant(-128, dtype=tf.int8)
maxval = tf.constant(127, dtype=tf.int8)

res = secure_random_module.secure_seeded_random_uniform(
    shape,
    seed,
    minval,
    maxval,
    name=None,
)
print("res = ", res)  # Output: [-128, -128]

```

## Impact:
This vulnerability directly affects secure comparison protocols in aby3,  making them insecure. Below is a call stack showing how this vulnerability can be triggered:

## Call Stack:

```

File "~/tf-encrypted/./test_rss_eq.py", line 142, in <module>
    test(args.secret, 1)
  File "~/tf-encrypted/./test_rss_eq.py", line 128, in test
    result = prot.equal(x, 0)
  File "~/tf-encrypted/tf_encrypted/protocol/protocol.py", line 117, in cache_nodes
    result = func(self, *args, **kwargs)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 1684, in equal
    return self.dispatch("equal", x, y)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 2038, in dispatch
    return func(self, *args, **kwargs)  # pylint: disable=not-callable
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 3053, in _equal_private_public
    return _equal_zero_private(prot, x - y)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 3064, in _equal_zero_private
    msb = prot.msb(pack)
  File "~/tf-encrypted/tf_encrypted/protocol/protocol.py", line 117, in cache_nodes
    result = func(self, *args, **kwargs)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 1561, in msb
    return self.dispatch("msb", x)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 2038, in dispatch
    return func(self, *args, **kwargs)  # pylint: disable=not-callable
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 3920, in _msb_private
    return _bit_extract_private(prot, x, x.backing_dtype.nbits - 1)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 3908, in _bit_extract_private
    carry = prot.carry(operand1, operand2, pos=i - 1)
  File "~/tf-encrypted/tf_encrypted/protocol/protocol.py", line 117, in cache_nodes
    result = func(self, *args, **kwargs)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 1724, in carry
    return self.dispatch("carry", x, y, pos)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 2038, in dispatch
    return func(self, *args, **kwargs)  # pylint: disable=not-callable
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 3579, in _carry_private_private
    return _carry_computation(prot, x, y, pos)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 3611, in _carry_computation
    P = Ps[0] & Ps[1]
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3_tensors.py", line 327, in __and__
    return self.bitwise_and(other)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3_tensors.py", line 320, in bitwise_and
    return self.prot.and_(self, other)
  File "~/tf-encrypted/tf_encrypted/protocol/protocol.py", line 117, in cache_nodes
    result = func(self, *args, **kwargs)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 1488, in and_
    return self.dispatch("and", x, y)
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 2038, in dispatch
    return func(self, *args, **kwargs)  # pylint: disable=not-callable
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 3161, in _and_private_private
    a0, a1, a2 = prot._gen_zero_sharing(
  File "~/tf-encrypted/tf_encrypted/protocol/aby3/aby3.py", line 957, in _gen_zero_sharing
    f00 = factory.sample_seeded_uniform(
  File "~/tf-encrypted/tf_encrypted/operations/secure_random/secure_random.py", line 109, in seeded_random_uniform
    res = secure_random_module.secure_seeded_random_uniform(
```

Let me know if further information is needed or if I can assist in testing a patch.

## Credit
Guopeng Lin ([Fudan University](https://dsg.fudan.edu.cn/#/))

**Comment by zicofish** [comment]
@GuopengLin Thanks for the issue. It is fixed in #901. Hasn't been merged, but please apply the fix and let us know whether it works.

**Comment by GuopengLin** [comment]
Thanks, it is fixed.
